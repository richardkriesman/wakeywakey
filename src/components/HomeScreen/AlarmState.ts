import { Alarm } from "../../models/Alarm";
import { AlarmEvent, AlarmEventType } from "../../services/TimerService";

export enum AlarmStateType { NONE, SLEEP, WAKE, GET_UP }

interface IAlarmState {
    type: AlarmStateType;
    alarm?: Alarm;
}

export class AlarmState implements IAlarmState {
    public static fromAlarmEvent(event: AlarmEvent): AlarmState {
        let type: AlarmStateType;

        switch (event.type) {
            case AlarmEventType.SLEEP:
                type = AlarmStateType.SLEEP;
                break;
            case AlarmEventType.WAKE:
                type = AlarmStateType.WAKE;
                break;
            case AlarmEventType.GET_UP:
                type = AlarmStateType.GET_UP;
                break;
            default:
                type = AlarmStateType.NONE;
                break;
        }

        return new AlarmState(type, event.alarm);
    }

    public static getMessageText(event: AlarmState): string {
        return messageTextMap.get(event.type);
    }

    public type: AlarmStateType;
    public alarm?: Alarm;

    public constructor(type: AlarmStateType, alarm?: Alarm) {
        this.type = type;
        this.alarm = alarm;
    }

    public toString(): string {
        return `{ type: ${this.type}, alarm_id: ${this.alarm.id} }`;
    }
}

const messageTextMap: Map<AlarmStateType, string> = new Map<AlarmStateType, string>([
    [AlarmStateType.NONE, "Hello, world!"],
    [AlarmStateType.SLEEP, "Time for bed!"],
    [AlarmStateType.WAKE, "Time to wake up!"],
    [AlarmStateType.GET_UP, "Time to get up!"]
]);
