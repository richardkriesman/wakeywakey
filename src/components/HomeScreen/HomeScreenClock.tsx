import React, { ReactNode } from "react";
import { View, StyleSheet, ViewStyle } from "react-native";

import { Text } from "react-native";

export interface HomeScreenClockProps {
    wrapperStyle?   : ViewStyle;
}

/**
 * The clock on the home screen.
 * 
 * @author Shawn Lutch
 */
export class HomeScreenClock extends React.Component<HomeScreenClockProps> {
    constructor(props : HomeScreenClockProps) {
        super(props);
    }

    _pad(num : number) : string {
        return (num < 10 ? '0' : '') + num;
    }

    render() : ReactNode {
        let now = new Date();

        return (
            <View style={this.props.wrapperStyle}>
                <Text style={styles.clock}>
                    {this._pad(now.getHours())}:{this._pad(now.getMinutes())}
                </Text>
            </View>
        );
    }
};

const styles = StyleSheet.create({
    clock: {
        color: 'black',
        fontSize: 70,
        marginTop: 20,
        marginBottom: 20,
        textAlign: 'center'
    }
});
