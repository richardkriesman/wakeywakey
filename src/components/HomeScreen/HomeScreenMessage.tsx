/**
 * @module components
 */

import React, { ReactNode } from "react";
import { StyleSheet, Text } from "react-native";

export interface HomeScreenMessageProps {
    text: string;
}

export class HomeScreenMessage extends React.Component<HomeScreenMessageProps> {
    public constructor(props: HomeScreenMessageProps) {
        super(props);
    }

    public render(): ReactNode {
        return (
            <Text style={styles.message}>{this.props.text}</Text>
        );
    }
}

const styles = StyleSheet.create({
    message: {
        fontSize: 30,
        textAlign: "center"
    }
});
