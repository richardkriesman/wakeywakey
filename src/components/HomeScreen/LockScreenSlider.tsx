/**
 * @module components
 */

import React, { Component, ReactNode } from "react";
import { ScrollView, StyleSheet, Text, View } from "react-native";

import VerticalSwipe from "react-native-vertical-swipe";

/**
 * The vertical slider for password on the home screen.
 *
 * @author Cody Kyrk
 */

export class LockScreenSlider extends Component {
    public render(): ReactNode {
        return (
            <VerticalSwipe
                offsetTop={300}   // Distance from top slider reaches
                closeSwipeThreshold={25} // Distance for slide to register
                openSwipeOffset={150} // Size of area open slide is registered
                // style={styles.dragContainer}
                content={(
                    <View style={styles.innerContainer}>
                        <ScrollView>
                            <Text style={styles.innerText}>
                                Spongebob me boy, enter that password! Arghegegegegegh
                            </Text>
                            <Text style={styles.innerText}>
                                Spongebob me boy, enter that password! Arghegegegegegh
                            </Text>
                        </ScrollView>
                    </View>
                )}>
            </VerticalSwipe>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        backgroundColor: "blue"
    },

    // Screen
    dragContainer: {
        alignItems: "center",
        backgroundColor: "gray",
        flex: 1,
        justifyContent: "center",
        zIndex: 2,
    },

    // Slider
    innerContainer: {
        alignSelf: "center",
        backgroundColor: "black",
        borderTopLeftRadius: 25,
        borderTopRightRadius: 25,
        height: 600,
        width: 300,
        zIndex: 2,
    },

    // Text in slider
    innerText: {
        color: "white",
        padding: 50
    }
});
