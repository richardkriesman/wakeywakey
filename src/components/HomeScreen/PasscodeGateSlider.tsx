/**
 * @module components
 */

import React, { ReactNode } from "react";
import {GestureResponderEvent, LayoutChangeEvent, StyleSheet, View, ViewStyle} from "react-native";

import { Button, Icon } from "react-native-elements";
import Colors from "../../constants/Colors";

interface PasscodeGateSliderState {
    y?: number;
}

export interface PasscodeGateSliderProps {
    initialTop: number;
    onIndicatorLayout?: (event: LayoutChangeEvent) => void;
}

// TODO: implement gravity, tapping to expand/contract, and rotating the arrow

export class PasscodeGateSlider extends React.Component<PasscodeGateSliderProps, PasscodeGateSliderState> {

    private contentHeight: number = 0;
    private yOffset: number = 0;

    public constructor(props: PasscodeGateSliderProps) {
        super(props);
        this.state = {};
    }

    public render(): ReactNode {

        // create dynamic style for parent view, allowing us to set the x and y coords
        const posStyle: ViewStyle = {
            top: this.y
        };

        return (
            <View
                onStartShouldSetResponder={() => true}
                onStartShouldSetResponderCapture={() => true}
                onResponderGrant={this.onResponderGrant.bind(this)}
                onResponderMove={this.onResponderMove.bind(this)}
                style={[posStyle, styles.container]}>
                <Button
                    buttonStyle={styles.indicator}
                    onLayout={this.onIndicatorLayout.bind(this)}
                    onPress={() => { return; }}
                    icon={
                        <Icon
                            name="up"
                            type="antdesign"
                            size={32}
                            color="white"
                        />
                    }
                />
                <View
                    onLayout={this.onContentLayout.bind(this)}
                    style={styles.content}>
                    {this.props.children}
                </View>
            </View>
        );
    }

    private onContentLayout(event: LayoutChangeEvent): void {
        this.contentHeight = event.nativeEvent.layout.height;
    }

    private onResponderGrant(event: GestureResponderEvent): void {
        this.yOffset = event.nativeEvent.pageY - this.y;
    }

    private onResponderMove(event: GestureResponderEvent): void {

        // compute new y
        let newY: number = event.nativeEvent.pageY - this.yOffset;
        if (newY > this.props.initialTop) { // y would exceed initial top, constrain y to initial top
            newY = this.props.initialTop;
        }
        if (newY < (this.props.initialTop - this.contentHeight)) {
            newY = this.props.initialTop - this.contentHeight;
        }

        // update y
        this.setState({
            y: newY
        });
    }

    private onIndicatorLayout(event: LayoutChangeEvent): void {
        if (this.props.onIndicatorLayout) {
            this.props.onIndicatorLayout(event);
        }
    }

    private get y(): number {
        return this.state.y ? this.state.y : this.props.initialTop;
    }
}

const styles = StyleSheet.create({
    container: {
        alignItems: "center",
        position: "absolute"
    },
    content: {

    },
    indicator: {
        backgroundColor: Colors.black,
        width: 200
    }
});
