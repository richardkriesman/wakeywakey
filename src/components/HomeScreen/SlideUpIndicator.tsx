import React, { ReactNode } from "react";
import { StyleSheet, View } from "react-native";

import { Button, Icon } from "react-native-elements";

/**
 * Properties that must be passed to <SlideUpIndicator />
 * @author Shawn Lutch
 */
export interface SlideUpIndicatorProps {
    // onPress event required
    onPress : VoidFunction;
}

/**
 * Indicator at the bottom of the screen to slide up.
 * @author Shawn Lutch
 */
export class SlideUpIndicator extends React.Component<SlideUpIndicatorProps> {

    protected onSwitch : VoidFunction;

    public constructor(props : SlideUpIndicatorProps) {
        super(props);
        this.onSwitch = props.onPress;
    }

    public render(): ReactNode {
        return (
            <Button
                buttonStyle={styles.indicator}
                onPress={this.onSwitch}
                icon={
                    <Icon
                        name="up"
                        type="antdesign"
                        size={32}
                        color="white"
                    />
                }
            />
        );
    }
};

const styles = StyleSheet.create({
    indicator: {
        backgroundColor: 'black',
        width: '100%'
    }
});
