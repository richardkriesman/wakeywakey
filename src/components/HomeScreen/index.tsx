export { HomeScreenClock as Clock } from "./HomeScreenClock";
export { SlideUpIndicator } from "./SlideUpIndicator";
export { SnoozeButton } from "./SnoozeButton";
