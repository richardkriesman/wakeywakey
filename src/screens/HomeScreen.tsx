import React, { ReactNode } from "react";
import { StyleSheet, View } from "react-native";
import { NavigationScreenProps } from "react-navigation";
import { Clock, Message, SlideUpIndicator, SnoozeButton } from "../components/HomeScreen";
import { NoHeader, UIScreen } from "../utils/screen";

/**
 * Home screen properties. Navigation by Miika, intersection type by Richard Kriesman.
 * @author Miika Raina, Richard Kriesman, Shawn Lutch
 */
export interface HomeScreenProps {
    initialMessageText: string;
}

/**
 * Home screen state
 * @author Richard Kriesman
 */
interface HomeScreenState {
    messageText: string;
}

/**
 * The home screen, where the current status, current time, and decorations will show.
 * @author Shawn Lutch, Miika Raina
 */
@NoHeader
export default class HomeScreen extends UIScreen<HomeScreenProps, HomeScreenState> {

    public static defaultInitialMessageText: string = "Hello, world!";

    public constructor(props: HomeScreenProps & NavigationScreenProps) {
        super(props);
    }

    public componentWillMount(): void {
        this.setState({ messageText: "Hello, world!" });
    }

    public renderContent(): ReactNode {
        return (
            <View style={ExtraStyles.container}>
                <View style={ExtraStyles.contentWrapper}>
                    <Message text={this.state.messageText} />
                    <Clock wrapperStyle={ExtraStyles.clockWrapper} />
                    <SnoozeButton onPress={this.onSnoozePressed.bind(this)} />
                </View>
                <View style={ExtraStyles.bottom}>
                    <SlideUpIndicator onPress={this.switchToSettings.bind(this)} />
                </View>
            </View>
        );
    }

    private switchToSettings(): void {
        this.setState({ messageText: "Switch to settings!" });
        this.present("SettingsMain");
    }

    private onSnoozePressed(): void {
        // TODO
        this.setState({ messageText: "Alarm snoozed!" });
    }

}

const ExtraStyles = StyleSheet.create({
    bottom: {
        alignItems: "center",
        bottom: -10,
        justifyContent: "center",
        position: "absolute",
        width: "100%",
        zIndex: 99
    },
    clockWrapper: {},
    container: {
        alignItems: "center",
        flex: 1,
        marginTop: 20,
        padding: 0
    },
    contentWrapper: {
        flex: 1,
        flexBasis: "100%",
        justifyContent: "center",
        width: "85%"
    }
});
