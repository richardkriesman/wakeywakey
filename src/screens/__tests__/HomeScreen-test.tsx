import React from "react";
import "react-native";
import renderer, { ReactTestInstance } from "react-test-renderer";
import { TestEnvironment } from "../../utils/testing";
import HomeScreen from "../HomeScreen";

const fakeDateMillis = 1551896555862;
const initialMessageText: string = "Lorem ipsum dolar sit amet";

let env: TestEnvironment;
beforeEach((done) => {
    TestEnvironment.init()
        .then((newEnv) => {
            env = newEnv;
            env.timing.date = fakeDateMillis;
            done();
        });
});

describe("app snapshot", () => {

    it("correctly fakes Date.now", async () => {
        const fakeNow = Date.now();
        expect(fakeNow).toBe(fakeDateMillis);
    });

    it("renders the screen", async () => {
        const tree = renderer.create(
            <HomeScreen initialMessageText={initialMessageText} {...env.emptyUIScreenProps} />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

});

describe("screen instance methods", () => {

    it("switches to settings", () => {
        // mock component
        const component = renderer.create(
            <HomeScreen initialMessageText={initialMessageText} {...env.emptyUIScreenProps}/>
        );

        // get instance
        const instance: (ReactTestInstance & HomeScreen) | null = component.getInstance() as any;
        expect(instance).toBeDefined();

        // spy on present
        const present = jest.spyOn(instance, "present");

        // switch to settings
        instance.switchToSettings();

        // check that everything was called correctly
        expect(present).toBeCalled();
    });

    it("snoozes the alarm", () => {
        // mock component
        const component = renderer.create(
            <HomeScreen initialMessageText={initialMessageText} {...env.emptyUIScreenProps}/>
        );

        // get instance
        const instance: (ReactTestInstance & HomeScreen) | null = component.getInstance() as any;
        expect(instance).toBeDefined();

        // spy on setState
        const setState = jest.spyOn(instance, "setState");

        // switch to settings
        instance.onSnoozePressed();

        // check that everything was called correctly
        expect(setState).toBeCalled();
    });

});
