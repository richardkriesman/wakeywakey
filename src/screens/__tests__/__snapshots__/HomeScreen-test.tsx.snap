// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app snapshot renders the screen 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "marginTop": 20,
            "padding": 0,
          }
        }
      >
        <View
          accessible={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text />
          <Text
            style={
              Object {
                "fontSize": 17,
              }
            }
          >
            Loading
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#5e5e5e",
                  "fontSize": 15,
                },
                undefined,
              ]
            }
          >
            Just a sec!
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "bottom": -10,
              "justifyContent": "center",
              "position": "absolute",
              "width": "100%",
              "zIndex": 99,
            }
          }
        >
          <View
            style={
              Object {
                "borderRadius": 3,
              }
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#000000",
                    "borderColor": "#2089dc",
                    "borderRadius": 3,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                    "width": "100%",
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                    theme={
                      Object {
                        "colors": Object {
                          "disabled": "hsl(208, 8%, 90%)",
                          "divider": "#bcbbc1",
                          "error": "#ff190c",
                          "grey0": "#393e42",
                          "grey1": "#43484d",
                          "grey2": "#5e6977",
                          "grey3": "#86939e",
                          "grey4": "#bdc6cf",
                          "grey5": "#e1e8ee",
                          "greyOutline": "#bbb",
                          "platform": Object {
                            "android": Object {
                              "error": "#f44336",
                              "primary": "#2196f3",
                              "secondary": "#9C27B0",
                              "success": "#4caf50",
                              "warning": "#ffeb3b",
                            },
                            "ios": Object {
                              "error": "#ff3b30",
                              "primary": "#007aff",
                              "secondary": "#5856d6",
                              "success": "#4cd964",
                              "warning": "#ffcc00",
                            },
                          },
                          "primary": "#2089dc",
                          "searchBg": "#303337",
                          "secondary": "#8F0CE8",
                          "success": "#52c41a",
                          "warning": "#faad14",
                        },
                      }
                    }
                    underlayColor="white"
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
